File: src/scandoc/tui/state.py
"""
Reactive Application State model for scanDOC TUI.
"""

from dataclasses import dataclass, field
from pathlib import Path
import os
from typing import Any, Dict, List, Optional

from scandoc.acceleration.models import DeviceType, PrecisionMode
from scandoc.models.document import DocumentIR
from scandoc.pipelines.models import PipelineConfig


class ScreenType:
    HOME = "home"
    FILE_PICKER = "file_picker"
    FOLDER_PICKER = "folder_picker"
    PIPELINE_CONFIG = "pipeline_config"
    PROCESSING = "processing"
    DOCUMENT_INSPECTOR = "document_inspector"
    EXPORT = "export"
    MODEL_MANAGER = "model_manager"
    BENCHMARK = "benchmark"
    SERVER_MANAGER = "server_manager"
    SETTINGS = "settings"
    HELP = "help"
    COMMAND_PALETTE = "command_palette"


@dataclass
class TuiState:
    """State tree for scanDOC TUI interface."""
    current_screen: str = ScreenType.HOME
    previous_screen: Optional[str] = None
    
    # File Navigation & Selection State
    current_dir: Path = field(default_factory=lambda: Path.cwd())
    selected_paths: List[Path] = field(default_factory=list)
    search_query: str = ""
    extension_filter: Optional[str] = None
    
    # Active Processing Document State
    active_document_ir: Optional[DocumentIR] = None
    active_document_path: Optional[Path] = None
    processing_status: str = "idle"  # 'idle', 'processing', 'completed', 'failed', 'cancelled'
    progress_stage: str = "Ready"
    progress_pct: float = 0.0
    current_page: int = 0
    total_pages: int = 0
    processing_errors: List[str] = field(default_factory=list)
    processing_logs: List[str] = field(default_factory=list)
    
    # Pipeline & Device Config
    pipeline_config: PipelineConfig = field(default_factory=PipelineConfig)
    offline_mode: bool = field(default_factory=lambda: os.getenv("SCANDOC_OFFLINE", "0") == "1")
    device_type: DeviceType = DeviceType.CPU
    precision_mode: PrecisionMode = PrecisionMode.FP32
    
    # Export Options
    selected_export_format: str = "markdown"
    export_output_dir: Path = field(default_factory=lambda: Path.cwd() / "output")
    include_images: bool = True
    preserve_formulas: bool = True
    preserve_tables: bool = True
    
    # Server State
    server_running: bool = False
    server_host: str = "127.0.0.1"
    server_port: int = 8000
    
    # Recent Documents History
    recent_documents: List[Dict[str, Any]] = field(default_factory=list)

    def is_offline(self) -> bool:
        return self.offline_mode or os.getenv("SCANDOC_OFFLINE", "0") == "1"

    def toggle_offline_mode(self) -> bool:
        self.offline_mode = not self.offline_mode
        os.environ["SCANDOC_OFFLINE"] = "1" if self.offline_mode else "0"
        return self.offline_mode

    def add_log(self, message: str) -> None:
        self.processing_logs.append(message)
        if len(self.processing_logs) > 500:
            self.processing_logs.pop(0)

    def add_recent(self, path: Path, status: str = "completed") -> None:
        rec = {
            "name": path.name,
            "path": str(path.resolve()),
            "status": status,
            "size_bytes": path.stat().st_size if path.exists() else 0,
        }
        self.recent_documents = [r for r in self.recent_documents if r["path"] != rec["path"]]
        self.recent_documents.insert(0, rec)
        self.recent_documents = self.recent_documents[:20]

File: internal/tui/state/state.go
package state

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// Screen identifiers matching Python ScreenType
const (
	ScreenHome              = "home"
	ScreenFilePicker        = "file_picker"
	ScreenOutputs           = "outputs"
	ScreenFolderPicker      = "folder_picker"
	ScreenPipelineConfig    = "pipeline_config"
	ScreenProcessing        = "processing"
	ScreenDocumentInspector = "document_inspector"
	ScreenExport            = "export"
	ScreenModelManager      = "model_manager"
	ScreenBenchmark         = "benchmark"
	ScreenServerManager     = "server_manager"
	ScreenSettings          = "settings"
	ScreenHelp              = "help"
	ScreenCommandPalette    = "command_palette"
)

type RecentDocument struct {
	Name      string    `json:"name"`
	Path      string    `json:"path"`
	Status    string    `json:"status"`
	SizeBytes int64     `json:"size_bytes"`
	Timestamp time.Time `json:"timestamp"`
}

type PipelineConfig struct {
	EnableOCR         bool   `json:"enable_ocr"`
	EnableLayout      bool   `json:"enable_layout"`
	EnableTable       bool   `json:"enable_table"`
	EnableFormula     bool   `json:"enable_formula"`
	EnableVLM         bool   `json:"enable_vlm"`
	RoutingMode       string `json:"routing_mode"` // "adaptive", "fast", "deep"
	FastModel         string `json:"fast_model"`
	DeepModel         string `json:"deep_model"`
}

type AppState struct {
	mu sync.RWMutex

	CurrentScreen  string
	PreviousScreen string

	// Navigation & File Selection
	CurrentDir      string
	WorkspaceDir    string
	WorkspaceRoot   string
	SelectedPaths   []string
	WindowWidth     int
	WindowHeight    int
	SearchQuery     string
	TickCount       int
	ExtensionFilter string

	// Active Document & Processing State
	ActiveDocumentPath string
	ActiveDocumentName string
	ProcessingStatus   string // "idle", "processing", "completed", "failed", "cancelled"
	ProgressStage      string
	ProgressPct        float64
	CurrentPage        int
	TotalPages         int
	ProcessingErrors   []string
	ProcessingLogs     []string

	// Pipeline & Device Config
	PipelineConfig PipelineConfig
	OfflineMode    bool
	DeviceType     string // "cpu", "cuda", "openvino"
	PrecisionMode  string // "fp32", "fp16", "int8"

	// Export Configuration
	ExportFormat    string
	ExportOutputDir string
	IncludeImages   bool
	PreserveFormulas bool
	PreserveTables   bool

	// Server Control State
	ServerRunning bool
	ServerHost    string
	ServerPort    int

	// Recent Documents History
	RecentDocuments []RecentDocument
}

func NewAppState() *AppState {
	cwd, err := os.Getwd()
	if err != nil {
		cwd = "."
	}
	return &AppState{
		CurrentScreen:   ScreenHome,
		CurrentDir:      cwd,
		WorkspaceDir:    filepath.Join(cwd, "local", "scandoc", "output"),
		WorkspaceRoot:   filepath.Join(cwd, "local", "scandoc", "output"),
		SelectedPaths:   make([]string, 0),
		ProcessingLogs:  make([]string, 0),
		ProcessingErrors: make([]string, 0),
		PipelineConfig: PipelineConfig{
			EnableOCR:         true,
			EnableLayout:      true,
			EnableTable:       true,
			EnableFormula:     true,
			EnableVLM:         true,
			RoutingMode:       "adaptive",
			FastModel:         "RapidOCR Mobile PP-OCRv4",
			DeepModel:         "RT-DETR DocLayNet",
		},
		OfflineMode:     os.Getenv("SCANDOC_OFFLINE") == "1",
		DeviceType:      "cpu",
		PrecisionMode:   "fp32",
		ExportFormat:    "markdown",
		ExportOutputDir: filepath.Join(cwd, "output"),
		IncludeImages:   true,
		PreserveFormulas: true,
		PreserveTables:   true,
		ServerHost:      "127.0.0.1",
		ServerPort:      8000,
		RecentDocuments: make([]RecentDocument, 0),
	}
}

func (s *AppState) IsOffline() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.OfflineMode || os.Getenv("SCANDOC_OFFLINE") == "1"
}

func (s *AppState) ToggleOfflineMode() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.OfflineMode = !s.OfflineMode
	if s.OfflineMode {
		os.Setenv("SCANDOC_OFFLINE", "1")
	} else {
		os.Setenv("SCANDOC_OFFLINE", "0")
	}
	return s.OfflineMode
}

func (s *AppState) NavigateTo(screen string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.PreviousScreen = s.CurrentScreen
	s.CurrentScreen = screen
}

func (s *AppState) NavigateBack() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.PreviousScreen != "" {
		s.CurrentScreen, s.PreviousScreen = s.PreviousScreen, ""
	} else {
		s.CurrentScreen = ScreenHome
	}
}

func (s *AppState) AddLog(msg string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.ProcessingLogs = append(s.ProcessingLogs, msg)
	if len(s.ProcessingLogs) > 500 {
		s.ProcessingLogs = s.ProcessingLogs[1:]
	}
}

func (s *AppState) AddRecent(path string, status string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	fi, err := os.Stat(path)
	var size int64 = 0
	if err == nil {
		size = fi.Size()
	}

	cleanPath := filepath.Clean(path)
	name := filepath.Base(cleanPath)

	rec := RecentDocument{
		Name:      name,
		Path:      cleanPath,
		Status:    status,
		SizeBytes: size,
		Timestamp: time.Now(),
	}

	filtered := make([]RecentDocument, 0, len(s.RecentDocuments))
	for _, r := range s.RecentDocuments {
		if !strings.EqualFold(r.Path, cleanPath) {
			filtered = append(filtered, r)
		}
	}

	s.RecentDocuments = append([]RecentDocument{rec}, filtered...)
	if len(s.RecentDocuments) > 20 {
		s.RecentDocuments = s.RecentDocuments[:20]
	}
}
